<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Stores the data and tutorials associated with the package `countsplit`  • countsplit.tutorials</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Stores the data and tutorials associated with the package `countsplit` ">
<meta property="og:description" content="What the package does (one paragraph).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">countsplit.tutorials</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/countsplit_tutorial.html">Introductory differential expression tutorial</a>
</li>
<li>
  <a href="articles/seurat_tutorial.html">Seurat tutorial</a>
</li>
<li>
  <a href="articles/scran_tutorial.html">Scran tutorial</a>
</li>
<li>
  <a href="articles/monocle3_tutorial.html">Monocle3 tutorial</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="welcome-to-the-tutorial-website-for-the-package-countsplit">Welcome to the tutorial website for the package <code>countsplit</code>
<a class="anchor" aria-label="anchor" href="#welcome-to-the-tutorial-website-for-the-package-countsplit"></a>
</h2>
<p>For the main package website, please visit <a href="https://anna-neufeld.github.io/countsplit/" class="external-link">https://anna-neufeld.github.io/countsplit/</a>.</p>
<p>The <code>countsplit</code> R package splits an integer-valued matrix into two or more folds of data that can be used for cross validation or for inference after latent variable estimation. If the entries in the original matrix follow a Poisson or a negative binomial distribution with known overdispersion parameter, then the folds of data will be independent and the inference or validation will be valid. The ability to handle negative binomial data and more than two folds of data is a new feature of the <code>countsplit</code> package; please read the new documentation carefully if you are an existing user.</p>
<p>The motivation for this method is described in Neufeld et al., 2022 <a href="http://arxiv.org/abs/2207.00554" class="external-link">(link to paper)</a> in the context of inference after latent variable estimation for Poisson distributed single cell RNA sequencing data. Briefly, count splitting allows users to perform differential expression analysis to see which genes vary across estimated cell types (such as those obtained via clustering) or along an estimated cellular trajectory (pseudotime). Additional motivation, along with the extension to negative binomial data, is described in Neufeld et al., 2023 <a href="...">(link to paper coming soon)</a>.</p>
<p>The tutorials on this website currently focus on the Poisson setting and on the inference after latent variable estimation setting, but they are being updated to include more negative binomial examples and more cross validation examples.</p>
</div>
<div class="section level2">
<h2 id="how-can-i-get-countsplit">How can I get countsplit?<a class="anchor" aria-label="anchor" href="#how-can-i-get-countsplit"></a>
</h2>
<p>Make sure that <code>remotes</code> is installed by running <code>install.packages("remotes")</code>, then type</p>
<p><code>{r} remotes::install_github("anna-neufeld/countsplit")</code></p>
<p>To also download the data needed to reproduce the package vignettes, be sure to also install the ``countsplit.tutorials” package.</p>
<p><code>{r} remotes::install_github("anna-neufeld/countsplit.tutorials").</code></p>
</div>
<div class="section level2">
<h2 id="where-can-i-learn-more">Where can I learn more?<a class="anchor" aria-label="anchor" href="#where-can-i-learn-more"></a>
</h2>
<p>See the <a href="articles/countsplit_tutorial.html">introductory tutorial</a> tab for an introduction to our framework on simple simulated data. See the <a href="articles/seurat_tutorial.html">seurat</a>, <a href="articles/scran_tutorial.html">scran</a>, and <a href="articles/monocle3_tutorial.html">monocle3</a> tutorials for examples of how the count splitting package can be integrated with common scRNA-seq analysis pipelines.</p>
<p>Please see the <a href="articles/demonstrating_problem.html">double dipping demonstration</a> for the code that goes with Appendix A of Neufeld et al., 2022.</p>
<p>Please visit <a href="https://github.com/anna-neufeld/countsplit_paper" class="external-link">https://github.com/anna-neufeld/countsplit_paper</a> for code to reproduce the figures and tables from Neufeld et al., 2022. Please visit <a href="https://github.com/anna-neufeld/nbcs_paper_simulations" class="external-link">https://github.com/anna-neufeld/nbcs_paper_simulations</a> for code to reproduce the figures and tables from Neufeld et al., 2023.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Neufeld, A.,Gao, L., Popp, J., Battle, A. &amp; Witten, D. (2022), ‘Inference after latent variable estimation for single-cell RNA sequencing data’, Biostatistics.</p>
<p>Neufeld, A.,Dharamshi, A., Gao, L., &amp; Witten, D. (2023), ‘Data thinning for convolution-closed distributions’, <a href="https://arxiv.org/abs/2301.07276/" class="external-link uri">https://arxiv.org/abs/2301.07276/</a> .</p>
<p>Neufeld, A., Popp, J., Gao, L., Battle, A. &amp; Witten, D. (2023), ‘Negative binomial count splitting for single-cell RNA sequencing data’.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing countsplit.tutorials</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Anna Neufeld <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
