<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial: differential expression analysis on single cell RNA-seq data • countsplit.tutorials</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial: differential expression analysis on single cell RNA-seq data">
<meta property="og:description" content="countsplit.tutorials">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">countsplit.tutorials</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/countsplit_tutorial.html">Differential expression tutorial</a>
</li>
<li>
  <a href="../articles/MSE_tutorial.html">Cross validation tutorial</a>
</li>
<li>
  <a href="../articles/seurat_tutorial.html">Seurat tutorial</a>
</li>
<li>
  <a href="../articles/scran_tutorial.html">Scran tutorial</a>
</li>
<li>
  <a href="../articles/monocle3_tutorial.html">Monocle3 tutorial</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial: differential expression analysis on
single cell RNA-seq data</h1>
            
      
      
      <div class="hidden name"><code>countsplit_tutorial.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we use two simple simulated data sets to
demonstrate how to use the countsplit package and base R functions
(e.g. glm, kmeans) to cluster and test for differential expression in
scRNA-seq data. See our [cross-validation tutorial]((<a href="https://anna-neufeld.github.io/countsplit.tutorials/articles/MSE_tutorial.html" class="external-link uri">https://anna-neufeld.github.io/countsplit.tutorials/articles/MSE_tutorial.html</a>)
for details on how to use the countsplit package to evaluate the output
of a clustering. See the non-introductory tutorials on this website to
learn how to integrate the countsplit package into three existing
scRNA-seq data analysis workflows for differential expression analysis
in R: <code>Seurat</code>, <code>scran</code>, and
<code>monocle3</code>. . We start by loading the packages we will be
working with. Make sure that <code>remotes</code> is installed by
running <code>install.packages("remotes")</code>, then type</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"anna-neufeld/countsplit"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/anna-neufeld/countsplit" class="external-link">countsplit</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="simulated-poisson-data-with-no-true-signal-">Simulated Poisson data with no true signal.<a class="anchor" aria-label="anchor" href="#simulated-poisson-data-with-no-true-signal-"></a>
</h2>
<p>First suppose that we have <span class="math inline">\(n=1000\)</span> cells and <span class="math inline">\(p=200\)</span> genes. Suppose that every count
<span class="math inline">\(\textbf{X}_{ij}\)</span> is drawn from a
<span class="math inline">\(\text{Poisson}(5)\)</span> distribution. We
first generate this data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">p</span>, lambda<span class="op">=</span><span class="fl">5</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">n</span><span class="op">)</span></span></code></pre></div>
<p>In this tutorial, we are interested in knowing which genes are
differentially expressed across discrete cell types. We will cluster
this data to estimate cell types and then test for differential
expression. Since there are no true cell types and no truly
differentially expressed genes in this data, a test of differential
expression that controls the Type 1 error rate should give uniformly
distributed p-values on this data.</p>
<div class="section level3">
<h3 id="the-naive-method">The naive method<a class="anchor" aria-label="anchor" href="#the-naive-method"></a>
</h3>
<p>First, we will demonstrate that estimating the clusters and testing
for differential expression using the same data does not control the
Type 1 error rate. We refer to the practice of using the same data for
clustering and differential expression testing as the “naive method” or
“double dipping”.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters.full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">X</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">X</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.full</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results.naive</span><span class="op">)</span></span>
<span><span class="co">##         Estimate Std. Error   z value     Pr(&gt;|z|)</span></span>
<span><span class="co">## [1,]  0.03958456 0.02820810  1.403305 1.605260e-01</span></span>
<span><span class="co">## [2,] -0.10237171 0.02857552 -3.582497 3.403252e-04</span></span>
<span><span class="co">## [3,]  0.11308989 0.02849956  3.968127 7.243959e-05</span></span>
<span><span class="co">## [4,]  0.10833419 0.02828266  3.830410 1.279301e-04</span></span>
<span><span class="co">## [5,] -0.08088836 0.02823097 -2.865235 4.166997e-03</span></span>
<span><span class="co">## [6,]  0.04539730 0.02797837  1.622586 1.046780e-01</span></span></code></pre></div>
<p>The first line clusters the cells by applying k-means clustering with
k=2 on the log-transformed data with a pseudocount of 1, and saves the
cluster assignments as <code>clusters.full</code>. The second line tests
for differential gene expression using Poisson GLMs. For every gene
<span class="math inline">\(X_j\)</span> in X, we fit a Poisson GLM of
<span class="math inline">\(X_j\)</span> on <code>clusters.full</code>,
and save the summary of the slope coefficient in
<code>results.naive</code>. As shown in the output, we have saved a
slope coefficient estimate, a standard error, a z-value, and a p-value
for every gene in the dataset.</p>
<p>Even in these first 6 rows of results, we can see that the naive
method assigns small p-values to many genes, despite the fact that no
genes are truly differentially expressed in this data. We can make a
uniform QQ-plot of the p-values for the naive method to see that they
are not uniformly distributed and thus do not control the Type 1
error.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.naive</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span>distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="poisson-count-splitting">Poisson count splitting<a class="anchor" aria-label="anchor" href="#poisson-count-splitting"></a>
</h3>
<p>We now address the issue using count splitting. In this section, we
will (correctly) assume that the data follow a Poisson distribution. In
the next section, we will consider the case where the data actually
follow a negative binomial distribution.</p>
<p>The key steps are (1) running the <code>countsplit</code> function to
get <code>Xtrain</code> and <code>Xtest</code> and then (2) clustering
the data using <code>Xtrain</code>, and then fitting a GLM to test for
association between those clusters and each column of
<code>Xtest</code>. The <code>countsplit</code> function returns a list,
which we call <code>split</code> here, which contains the training set
and the test set.</p>
<p>Since we only want a single training set and a single test set, we
will set <code>folds=2</code> when calling <code>countsplit</code>. By
default, this creates two identically distributed folds that each store
half of the information in the dataset. We can change the amount of
information allocated to the two folds by changing the
<code>epsilion</code> parameter. When <code>epsilon</code> is very close
to 0, the training matrix will be extremely sparse and the test matrix
will look a lot like <code>X</code>. When epsilon is very close to 1,
the training matrix will be nearly identical to <code>X</code>, but the
test matrix will be extremely sparse. The default in the
<code>countsplit</code> function is to set
<code>epsilon=c(0.5, 0.5)</code>. See <a href="https://arxiv.org/abs/2207.00554" class="external-link">our preprint</a> for more
information.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit</a></span><span class="op">(</span><span class="va">X</span>, folds<span class="op">=</span><span class="fl">2</span>, epsilon<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Like before, we will cluster using k-means with log-transformed data
and we will fit Poisson GLMs for differential expression. Unlike before,
we will run the clustering on <code>Xtrain</code> and use
<code>Xtest</code> as the response in our GLMs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.countsplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results.countsplit</span><span class="op">)</span></span>
<span><span class="co">##          Estimate Std. Error    z value   Pr(&gt;|z|)</span></span>
<span><span class="co">## [1,]  0.028171056 0.03988386  0.7063272 0.47998469</span></span>
<span><span class="co">## [2,] -0.014424364 0.04075410 -0.3539365 0.72338648</span></span>
<span><span class="co">## [3,]  0.007947168 0.04093763  0.1941287 0.84607511</span></span>
<span><span class="co">## [4,]  0.024273036 0.04004166  0.6061945 0.54438561</span></span>
<span><span class="co">## [5,] -0.067879740 0.04045377 -1.6779582 0.09335526</span></span>
<span><span class="co">## [6,] -0.052881527 0.04004365 -1.3205971 0.18663572</span></span></code></pre></div>
<p>We can see from the summary output that the p-values for the first 6
genes are much larger. When we make the same uniform QQ-plot as before,
we see that the p-values obtained from count splitting are uniformly
distributed. Since no genes are differentially expressed in our data,
this means that count splitting controls the Type 1 error.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span>distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>The crucial property that ensures that count splitting controls the
Type 1 error in this setting is independence between <code>Xtrain</code>
and <code>Xtest</code>. We can see that the columns of
<code>Xtrain</code> are independent of the columns of <code>Xtest</code>
in this data with no signal by looking at the sample correlations, which
are centered around <span class="math inline">\(0\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">p</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">[</span>,<span class="va">u</span><span class="op">]</span>, <span class="va">Xtest</span><span class="op">[</span>,<span class="va">u</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">cors</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-10-1.png" width="700">
In summary, count splitting controls the Type 1 error when there is no
true signal in the data.</p>
</div>
</div>
<div class="section level2">
<h2 id="simulated-negative-binomial-data-with-no-true-signal">Simulated negative binomial data with no true signal<a class="anchor" aria-label="anchor" href="#simulated-negative-binomial-data-with-no-true-signal"></a>
</h2>
<p>We now generate data with no true signal, but we generate the data
from a negative binomial distribution. We use the mean + overdispersion
parameterization of the negative binomial, in which the mean is given by
<code>mu</code> and the variance is given by <code>mu+mu^2/5</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">p</span>, mu<span class="op">=</span><span class="fl">5</span>, size<span class="op">=</span><span class="fl">5</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">n</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Sanity check on parameterization</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] 4.99826</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] 9.966747</span></span>
<span><span class="fl">5</span><span class="op">+</span><span class="fl">5</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fl">5</span></span>
<span><span class="co">## [1] 10</span></span></code></pre></div>
<p>Once again, in this part of the tutorial, we will cluster the data to
estimate cell types and then test for differential expression. Since
there are no true cell types and no truly differentially expressed genes
in this data (all genes and all cells have mean <span class="math inline">\(5\)</span>), a test of differential expression
that controls the Type 1 error rate should give uniformly distributed
p-values on this data. We will compare the naive method to three version
of count splitting.</p>
<p>We first note that the naive method still fails to control the Type 1
error rate. We use a negative binomial GLM (as implemented in the
<code>MASS</code> package, rather than a Poisson GLM, throughout this
section).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters.full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">X</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">X</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/glm.nb.html" class="external-link">glm.nb</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.full</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.naive</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span>distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Unfortunately, we now also note that the default settings of count
splitting will cause us to fail to control the Type 1 error rate. Note
that we omit the arguments for <code>folds</code> and
<code>epsilon</code> , because two equally sized folds is the default
for count splitting.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.countsplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/glm.nb.html" class="external-link">glm.nb</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.naive</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Naive method"</span><span class="op">)</span>,</span>
<span>  distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Poisson count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>By default, the <code>countsplit</code> function assumes that your
data follow a Poisson distribution. Now that the data are not Poisson
distributed, the default settings of <code>countsplit</code> fail to
lead to independent training and test sets, as we can see by looking at
the sample correlations between columns of <code>Xtrain</code> and
columns of <code>Xtest</code>. These correlations will get worse as the
amount of overdispersion in the negative binomial data increases.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">p</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">[</span>,<span class="va">u</span><span class="op">]</span>, <span class="va">Xtest</span><span class="op">[</span>,<span class="va">u</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">cors</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>To deal with this, we must tell the <code>countsplit</code> function
that our data are overdispersed. In this toy setting, we know the true
value of the overdispersion parameter; it is <span class="math inline">\(5\)</span> for all genes. So we can perform the
ideal version of negative binomial count splitting where we plug in the
value of <span class="math inline">\(5\)</span> for every gene.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">splitNB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit</a></span><span class="op">(</span><span class="va">X</span>, overdisps<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">5</span>,<span class="va">p</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">splitNB</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">splitNB</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.countsplit.NB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/glm.nb.html" class="external-link">glm.nb</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.naive</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Naive method"</span><span class="op">)</span>,</span>
<span>  distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Poisson count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit.NB</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Ideal NB count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We see that we have recovered Type 1 error control. Unfortunately, in
practice, we typically will not know the true value of the
overdispersion parameter and will instead need to plug in a
gene-specific estimate. A common way to estimate gene-specific
overdispersion parameters is with the <code>sctransform</code> package
in <code>R</code>. More specifically, if we call the <code>vst()</code>
function from the <code>sctransform</code> package on the gene-by-cell
matrix of counts (the transpose of <span class="math inline">\(X\)</span>, in this case), the estimated
gene-specific overdispersion paramateters are stored in the first column
of the <code>model_pars</code> section of the output.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">rownames</span>(X) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span>n</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">colnames</span>(X) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span>p</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>overdisps.est <span class="ot">&lt;-</span> sctransform<span class="sc">::</span><span class="fu">vst</span>(<span class="fu">t</span>(X))<span class="sc">$</span>model_pars[,<span class="dv">1</span>]</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span></code></pre></div>
<p>By default, the <code>vst()</code> function does not necessarily
return an estimated overdispersion for every single gene in the dataset.
For example, it only returns an estimated overdispersion for genes that
were expressed in at least 5 cells, and that were not found to be
approximately Poisson distributed. The code below verifies that, in this
case, we did in fact recieve an estimated overdispersion for all
cells.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/all.equal-methods.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">overdisps.est</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] TRUE</span></span></code></pre></div>
<p>We can see that, in this case, the estimation of the overdispersion
parameters went fairly well. All of the true overdispersion parameters
are 5, and the estimated overdispersion parameters are all between <span class="math inline">\(4\)</span> and <span class="math inline">\(6\)</span>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">overdisps.est</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>We are now prepared to apply negative binomial count splitting with
these estimated parameters. We see that the performance is approximately
as good as it was in the ideal version of negative binomial count
splitting.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">splitNB.est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit</a></span><span class="op">(</span><span class="va">X</span>, overdisps<span class="op">=</span><span class="va">overdisps.est</span><span class="op">)</span></span>
<span><span class="va">Xtrain.est</span> <span class="op">&lt;-</span> <span class="va">splitNB.est</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Xtest.est</span> <span class="op">&lt;-</span> <span class="va">splitNB.est</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">clusters.train.est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Xtrain.est</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="va">results.countsplit.NB.est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest.est</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/glm.nb.html" class="external-link">glm.nb</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train.est</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.naive</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Naive method"</span><span class="op">)</span>,</span>
<span>  distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Poisson count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit.NB</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Ideal NB count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_qq.html" class="external-link">geom_qq</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>sample<span class="op">=</span><span class="va">results.countsplit.NB.est</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Estimated NB count splitting"</span><span class="op">)</span>, distribution<span class="op">=</span><span class="fu">stats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">qunif</a></span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="simulated-poisson-data-with-true-signal">Simulated Poisson data with true signal<a class="anchor" aria-label="anchor" href="#simulated-poisson-data-with-true-signal"></a>
</h2>
<p>We now return to the Poisson setting, for simplicity. We demonstrate
the performance of count splitting on a simple simulated dataset that
contains two true clusters. We first randomly assign the cells to one of
two true clusters. We then generate data such that <span class="math inline">\(X_{ij} \sim
\mathrm{Poisson}(\Lambda_{ij})\)</span>. Genes 1-10 are differentially
expressed– for <span class="math inline">\(j=1,\ldots,10\)</span>, <span class="math inline">\(\Lambda_{ij} = 5\)</span> for cells in cluster
<span class="math inline">\(0\)</span> and <span class="math inline">\(\Lambda_{ij}=10\)</span> for cells in cluster
<span class="math inline">\(1\)</span>. Genes 11-200 are not
differentially expressed (<span class="math inline">\(\Lambda_{ij}=5\)</span> for all cells).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">clusters.true</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">n</span>, size<span class="op">=</span><span class="fl">1</span>, prob<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">Lambda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">5</span>, nrow<span class="op">=</span><span class="va">n</span>, ncol<span class="op">=</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="va">Lambda</span><span class="op">[</span><span class="va">clusters.true</span><span class="op">==</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Lambda</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="va">rpois</span>,n<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We now count split the data and save <code>Xtrain</code> and
<code>Xtest</code> for later use. Note that we don’t actually need to
specify that we want two folds of data, as this is the default
setting.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">split</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">Xtrain</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Xtest</span> <span class="op">&lt;-</span> <span class="va">split</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="section level3">
<h3 id="effect-of-using-xtrain-for-cluster-estimation-">Effect of using Xtrain for cluster estimation.<a class="anchor" aria-label="anchor" href="#effect-of-using-xtrain-for-cluster-estimation-"></a>
</h3>
<p>First, let’s look at the effect of count splitting on our ability to
estimate the true clusters. If we use all of our data <code>X</code> to
estimate the clusters, we make only 5 errors.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters.full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">X</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clusters.true</span>, <span class="va">clusters.full</span><span class="op">)</span></span>
<span><span class="co">##              clusters.full</span></span>
<span><span class="co">## clusters.true   1   2</span></span>
<span><span class="co">##             0   5 515</span></span>
<span><span class="co">##             1 480   0</span></span></code></pre></div>
<p>If instead we use only <code>Xtrain</code> to estimate the clusters,
we make a few additional errors, but we still come very close to
estimating the true clusters.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Xtrain</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">clusters.true</span>, <span class="va">clusters.train</span><span class="op">)</span></span>
<span><span class="co">##              clusters.train</span></span>
<span><span class="co">## clusters.true   1   2</span></span>
<span><span class="co">##             0  17 503</span></span>
<span><span class="co">##             1 468  12</span></span></code></pre></div>
<p>The fact that cluster 0 in the true clustering maps to cluster 2 in
the training set clustering is unimportant, but we will remember this
later when we attempt to compare models.</p>
</div>
<div class="section level3">
<h3 id="effect-of-using-xtest-for-inference">Effect of using Xtest for inference<a class="anchor" aria-label="anchor" href="#effect-of-using-xtest-for-inference"></a>
</h3>
<p>We now compare what happens when we regress <span class="math inline">\(X_j\)</span> on the true clusters to what happens
if we regress <span class="math inline">\(X_j^{\mathrm{test}}\)</span>
on the true clusters. We use the code below to fit a Poisson GLM of
every gene on the true clusters; we save both the slope and the
intercept. We are interested in the GLM slopes, as the GLM slopes are
non-zero if the gene is differentially expressed across the
clusters.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coeffs.X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">X</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.true</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coeffs.Xtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.true</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The plot below shows that the slopes resulting from <code>X</code>
and the slopes resulting from <code>Xtest</code> tend to fall on the
diagonal line <code>y=x</code> and so tend to be approximately equal to
eachother. The intercepts, on the other hand, get shifted by
<code>log(0.5)</code> when we use the <code>Xtest</code> rather than
<code>X</code>. This is as we would expect (see Section 4.1 of <a href="https://arxiv.org/abs/2207.00554" class="external-link">our preprint</a>).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">differentially_expressed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">190</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">coeffs.X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, y<span class="op">=</span><span class="va">coeffs.Xtest</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, col<span class="op">=</span><span class="va">differentially_expressed</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span> <span class="fl">0</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Intercepts from X"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Intercepts from Xtest"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Intercepts"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">coeffs.X</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, y<span class="op">=</span><span class="va">coeffs.Xtest</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">differentially_expressed</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span><span class="fl">0</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>,<span class="fl">1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>,<span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Slopes from X"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Slopes from Xtest"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Slopes"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="overall-comparison-of-count-splitting-to-ideal-method">Overall comparison of count splitting to ideal method<a class="anchor" aria-label="anchor" href="#overall-comparison-of-count-splitting-to-ideal-method"></a>
</h3>
<p>Finally, we compare the overall slope and intercept estimates that we
get from count splitting to what we would get in the ideal setting where
we get to regress <span class="math inline">\(X_j\)</span> on the true
clusters. Note that <code>coeffs.ideal</code> is identical to
<code>coeffs.X</code> from the previous section; we reproduce it here
for convenience. Note also that we regress <code>Xtest</code> on
<code>as.factor(clusters.train==1)</code>, rather than
<code>as.factor(clusters.train)</code>, simply because we saw above that
the “cluster 1” in the training set maps to the “cluster 1” in the true
clustering, and this consistency will ensure that the coefficients from
the two models have the same sign.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coeffs.ideal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">X</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.true</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coeffs.countsplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Xtest</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">u</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">u</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">clusters.train</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span>, family<span class="op">=</span><span class="st">"poisson"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">coeffs.ideal</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, y<span class="op">=</span><span class="va">coeffs.countsplit</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, col<span class="op">=</span><span class="va">differentially_expressed</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span> <span class="fl">0</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Intercepts from ideal method"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Intercepts from count splitting"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Intercepts"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="cn">NULL</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">coeffs.ideal</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, y<span class="op">=</span><span class="va">coeffs.countsplit</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">differentially_expressed</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept<span class="op">=</span><span class="fl">0</span>, slope<span class="op">=</span><span class="fl">1</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>,<span class="fl">1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.15</span>,<span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Slopes from ideal method"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Slopes from count splitting"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Slopes"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides<span class="op">=</span><span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="countsplit_tutorial_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>Overall, we see general agreement between the parameters estimated
via countsplitting and those estimated via the ideal method. The slopes
tend to be the same, whereas the intercepts are shifted by
<code>log(0.5)</code>, as expected based on <a href="https://arxiv.org/abs/2207.00554" class="external-link">our preprint</a>. The slopes are
the quantities that we care to do inference on, as they measure
differential expression across clusters.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anna Neufeld.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
